src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.h

```diff
@@ -0,0 +1,1339 @@
+template<class T>
+size_t size_of(const T& t) {
+  using decayed_t = std::decay_t<T>;
+  if constexpr (std::is_scalar_v<decayed_t>) {
+    return sizeof(decayed_t);
+  } else {
+    return t.size();
+  }
+}
+
+enum class size_state_t {
+  okay,
+  underflow,
+  overflow,
+};
+
+// layout of a node
+template<size_t BlockSize,
+         int N,
+         ntype_t NodeType>
+struct node_t {
+  static_assert(N == std::clamp(N, 0, 3));
+  constexpr static ntype_t node_type = NodeType;
+  constexpr static int node_n = N;
+
+  using key_t = FixedKey<N, NodeType>;
+  using item_t = std::conditional_t<NodeType == ntype_t::leaf,
+                                    onode_t,
+                                    child_addr_t>;
+  using const_item_t = std::conditional_t<NodeType == ntype_t::leaf,
+                                          const onode_t&,
+                                          child_addr_t>;
+  // XXX: magic numbers here
+  static constexpr bool item_in_key = (N == 3 &&
+                                       NodeType == ntype_t::inner);
+  using partial_key_t = std::conditional_t<N < 3,
+                                           variable_sized_key_t,
+                                           empty_key_t>;
+
+  std::pair<const key_t&, const partial_key_t&>
+  key_at(unsigned slot) const {
+    auto& key = keys[slot];
+    if constexpr (item_in_key) {
+      return {key, partial_key_t{}};
+    } else {
+      auto p = from_end(key.offset);
+      return {key, *reinterpret_cast<const partial_key_t*>(p)};
+    }
+  }
+
+  // update an existing oid with the specified item
+  ghobject_t update_oid_with_slot(unsigned slot, const ghobject_t& oid) const {
+    auto [prefix, suffix] = key_at(slot);
+    ghobject_t updated = oid;
+    prefix.update_oid(updated);
+    suffix.update_oid(updated);
+    return updated;
+  }
+
+  const_item_t item_at(const key_t& key) const {
+    if constexpr (item_in_key) {
+      return key.child_addr;
+    } else {
+      assert(key.offset < BlockSize);
+      auto p = from_end(key.offset);
+      auto partial_key = reinterpret_cast<const partial_key_t*>(p);
+      p += size_of(*partial_key);
+      return *reinterpret_cast<const item_t*>(p);
+    }
+  }
+
+  void dump(std::ostream& os) const {
+    for (uint16_t i = 0; i < count; i++) {
+      const auto& [prefix, suffix] = key_at(i);
+      os << " [" << i << '/' << count - 1 << "]\n"
+         << "  key1 = (" << prefix << ")\n"
+         << "  key2 = (" << suffix << ")\n";
+      const auto& item = item_at(prefix);
+      if (_is_leaf()) {
+        os << " item = " << item << "\n";
+      } else {
+        os << " child = " << std::hex << item << std::dec << "\n";
+      }
+    }
+  }
+
+  // for debugging only.
+  static constexpr bool is_leaf() {
+    return node_type == ntype_t::leaf;
+  }
+
+  bool _is_leaf() const {
+    return tag.is_leaf();
+  }
+
+  char* from_end(uint16_t offset) {
+    auto end = reinterpret_cast<char*>(this) + BlockSize;
```

Comment:
`std::uintptr_t`?

Reply:
i am not sure if i want to go this way, the type of `offset` should match that of `offset` in `FixedKey` s which have offset. `std::uintptr_t` is more correct is i were defining a API to be consumed by external user.